cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -360,6 +361,15 @@ class SnappyUnifiedMemoryManager private[memory](
           return
         }
 
+        // Stop execution pool to grow beyond eviction percentage of heap.
+        if (memoryMode eq MemoryMode.ON_HEAP) {
+          if (executionPool.memoryUsed + extraMemoryNeeded > maxHeapExecutionSize) {
+            logWarning(s"MemoryManager can't allocate $numBytes bytes as it " +
+                s"would exceed maxHeapExecutionSize = $maxHeapExecutionSize")
+            return
+          }
+        }
+
```

Comment:
This is the case when storage has not reached till storageFraction. Should we not allow execution to borrow from storage in such cases?  We can stop execution pool to grow beyond storageFraction, but some space would be wasted if the user has allocated memory, but table data is less.
The check for ON-HEAP is done as critical heap kicks in if push heap memory too much. OFF-HEAP memory can take all of storage free memory.